### DISCUSSION TITLE
Key handle with sk-ssh-ed25519 keys

### DISCUSSION THREAD

**Discussion by filom** [libssh2/libssh2#1061] [Q&A] [answered]
I'm trying to integrate the new api to authenticate using the `sk-ssh-ed25519` but I have a problem with the key handle (I believe that is the problem). 
The process works with `sk-ecdsa-sha2-nistp256`, allowing the ssh authentication to complete successfully. 

Based on the U2F protocol https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-raw-message-formats-v1.2-ps-20170411.html  2 steps are required: first the U2F registration and then the U2F authentication. 
The registration return a key handle that apparently should be used on the second step, but the libssh2 api call back provide also a key handle which has the same length in case of `sk-ecdsa-sha2-nistp256` but different content, while in case of `sk-ssh-ed25519` is has different length and content.
So using the key handle from the ssh callback make the `sk-ecdsa-sha2-nistp256` authentication possible, while with `sk-ssh-ed25519` it fails (using either key handle, from u2f authentication or ssh callback).

I also notice that U2F spec says that signature result from the u2f authentication step is always ECDSA, so why the `_LIBSSH2_SK_SIG_INFO` supports also Ed25519 ? Is there any relation?

Thanks for the support!

Version:
- iOS 16.1
- libssh2 master branch (1.10.1-dev)
- crypto backend: OpenSSL

**Comment by MichaelBuckley** [comment, marked as answer]
@filom 

Hello,

libssh2's support for FIDO keys is based only on the following document. https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.u2f

The libssh2 callback, must fill in the sig_info struct, and this struct does indeed look like an ECDSA signature. However, when authenticating with ED25519, you can set the sig_s_len to 0 and place the ed25519 signature in sig_r.

On iOS, I only have experience using the YubiKit framework, specifically for YubiKey devices. On desktop, however, here is the general code you would use for libfido2 and OpenSSL to authenticate. I removed a lot of error handling to make it simpler. Hopefully, it should show the steps you need to take in whatever library you are using. Please let me know if you have further questions.

```fido_dev_info_t *list = fido_dev_info_new(maxDevices);
const fido_dev_info_t *device_info = fido_dev_info_ptr(list, i);
const char *path = fido_dev_info_path(device_info);
fido_dev_t *fido_dev_new();
fido_dev_open(device, path);
fido_assert_t *assert = fido_assert_new();
fido_assert_set_clientdata(assert, data, data_len);
fido_assert_set_rp(assert, application);
fido_assert_allow_cred(assert, key_handle, handle_len);
fido_assert_set_up(assert, (flags & LIBSSH2_SK_PRESENCE_REQUIRED) ? FIDO_OPT_TRUE : FIDO_OPT_FALSE);
fido_assert_set_uv(assert, FIDO_OPT_TRUE);
fido_dev_get_assert(device, assert, pin); // pin is NULL is not required
const unsigned char *sig_ptr = fido_assert_sig_ptr(assert, 0);
size_t sig_len = fido_assert_sig_len(assert, 0);
sig_info->flags = fido_assert_flags(assert, 0);
sig_info->counter = fido_assert_sigcount(assert, 0);

if (algorithm == LIBSSH2_HOSTKEY_TYPE_ED25519) {
    sig_info->sig_r_len = sig_len;
    sig_info->sig_r = LIBSSH2_ALLOC(session, sig_len);
    memcpy(sig_info->sig_r, sig_ptr, sig_len);
} else if ( algorithm == LIBSSH2_HOSTKEY_TYPE_ECDSA_256) {
    ECDSA_SIG *ecdsa_sig = d2i_ECDSA_SIG(NULL, &sig_ptr, sig_len);
    const BIGNUM *sig_r, *sig_s;
    ECDSA_SIG_get0(ecdsa_sig, &sig_r, &sig_s);
    sig_info->sig_r_len = BN_num_bytes(sig_r);
    sig_info->sig_s_len = BN_num_bytes(sig_s);
    sig_info->sig_r = LIBSSH2_ALLOC(session, sig_info->sig_r_len);
    sig_info->sig_s = LIBSSH2_ALLOC(session, sig_info->sig_s_len);
    BN_bn2bin(sig_r, sig_info->sig_r);
    BN_bn2bin(sig_s, sig_info->sig_s);
}

fido_dev_close(device);
fido_dev_free(&device);
fido_assert_free(&assert);
```

**Comment by filom** [comment]
Thanks for the example, I tested on macOS and it works correctly with both ecdsa and ed25519. 
My goal is to make it work on iOS with NFC, but in this scenario the actual signing using ed25519 seems rejected by the key. Probably the libfido do something different based on the key algorithm or is the nfc command that does not support ed25519.
